web3/_utils/abi.py

```diff
@@ -580,11 +675,28 @@ def abi_sub_tree(abi_type: Optional[Union[TypeStr, ABIType]], data_value: Any) -
     if isinstance(abi_type, TypeStr):
         abi_type = parse(abi_type)
 
+    # In the two special cases below, we rebuild the given data structures with
+    # annotated items
     if abi_type.is_array:
-        it = abi_type.item_type
-        return ABITypedData([abi_type.to_type_str(), [abi_sub_tree(it, i) for i in data_value]])
-    else:
-        return ABITypedData([abi_type.to_type_str(), data_value])
+        # If type is array, determine item type and annotate all
+        # items in iterable with that type
+        item_type_str = abi_type.item_type.to_type_str()
+        data_value = [
```

Comment:
Same here with convention of not overwriting local variables which if fixed will require this `if/elif` to get a final `else` clause.